### ISSUE TITLE
Orders filled but algorithm still considers them open

### ISSUE THREAD

**Issue by sam31415** [enigmampc/catalyst#309]
Hi,

See [this issue](https://github.com/enigmampc/catalyst/issues/307) for my environment.

# Description of Issue

When selling positions in bch_btc and ltc_btc on Poloniex using effectively market orders (actually limit orders with limit one quarter of the current price), the following happened (according to the log below):
- The two orders were placed at 06:01:00.
- They were filled around 6:01:02.
- A few seconds later, the algorithm somehow saw the orders as still open. It mentions that they have been open for 48 seconds, which doesn't make sense to me: they were placed a few seconds earlier. Or maybe this refers to when the ordering function was called?
- The algorithm detected that the positions in bch_btc and ltc_btc were now zero, and updated itself without adding to the cash the proceeds of the sale.
- This results in an apparent loss corresponding to the amount of the sold positions.
See the output of the algorithm below. The algorithm uses order_target() to do the ordering.

I should emphasize this behavior is rather random. Numerous positions were sold previously without problem. Maybe the problem lies in the way that Catalyst deals with imperfect network connections.

```
[2018-04-20 06:01:04.910546] INFO: exchange_blotter: poloniex order 205328425212 for etc_btc still open after 0 days 00:00:48.910343
[2018-04-20 06:01:06.870621] INFO: exchange_blotter: poloniex order 118498224337 for zec_btc still open after 0 days 00:00:48.870499
[2018-04-20 06:01:07.968919] WARNING: Exchange: detected lower balance for etc_btc on poloniex: 1e-08 < 1.21984671, updating position amount
[2018-04-20 06:01:07.969499] WARNING: Exchange: detected lower balance for zec_btc on poloniex: 0.0 < 0.08515753, updating position amount
```

Here is how you can reproduce this issue on your machine:

## Reproduction Steps

1. Let an algorithm buy and sell positions on Poloniex with market orders for some time. Sorry not to be able to be more specific.

Thanks,

Samuel

**Comment by sam31415** [comment]
The same problem occurred again today:
```
[2018-04-21 10:01:02.202519] INFO: exchange_blotter: poloniex order 444153990559 for eth_btc still open after 0 days 00:00:49.202378                           
[2018-04-21 10:01:04.188911] INFO: exchange_blotter: filled order 268872239733 / xmr_btc: -0.08798392, avg price: 0.0287                                       
[2018-04-21 10:01:04.387562] INFO: exchange_blotter: poloniex order 171699774378 for dash_btc still open after 0 days 00:00:46.387431                          
[2018-04-21 10:01:08.552261] WARNING: Exchange: detected lower balance for eth_btc on poloniex: 1e-08 < 0.03785018, updating position amount                   
[2018-04-21 10:01:08.552671] WARNING: Exchange: detected lower balance for dash_btc on poloniex: 0.0 < 0.05010401, updating position amount                    
[2018-04-21 10:02:01.740460] INFO: exchange_blotter: poloniex order 444153990559 for eth_btc still open after 0 days 00:01:48.740303                           
[2018-04-21 10:02:02.063831] INFO: exchange_blotter: poloniex order 171699774378 for dash_btc still open after 0 days 00:01:44.063704                          
[2018-04-21 10:03:01.040884] INFO: exchange_blotter: poloniex order 444153990559 for eth_btc still open after 0 days 00:02:48.040747                           
[2018-04-21 10:03:03.544402] INFO: exchange_blotter: poloniex order 171699774378 for dash_btc still open after 0 days 00:02:45.544278                          
[2018-04-21 10:04:00.168558] INFO: exchange_blotter: poloniex order 444153990559 for eth_btc still open after 0 days 00:03:47.168406                           
[2018-04-21 10:04:01.171477] INFO: exchange_blotter: poloniex order 171699774378 for dash_btc still open after 0 days 00:03:43.171325                          
[2018-04-21 10:05:00.986617] INFO: exchange_blotter: poloniex order 444153990559 for eth_btc still open after 0 days 00:04:47.986476                           
[2018-04-21 10:05:01.071377] INFO: exchange_blotter: poloniex order 171699774378 for dash_btc still open after 0 days 00:04:43.071244
[2018-04-21 10:06:01.367336] INFO: exchange_blotter: poloniex order 444153990559 for eth_btc still open after 0 days 00:05:48.367172                           
[2018-04-21 10:06:03.006136] INFO: exchange_blotter: poloniex order 171699774378 for dash_btc still open after 0 days 00:05:45.006006                          
[2018-04-21 10:07:02.250078] INFO: exchange_blotter: poloniex order 444153990559 for eth_btc still open after 0 days 00:06:49.249934                           
[2018-04-21 10:07:02.823603] INFO: exchange_blotter: poloniex order 171699774378 for dash_btc still open after 0 days 00:06:44.823465                          
[2018-04-21 10:08:00.807984] INFO: exchange_blotter: poloniex order 444153990559 for eth_btc still open after 0 days 00:07:47.807832                           
[2018-04-21 10:08:01.780877] INFO: exchange_blotter: poloniex order 171699774378 for dash_btc still open after 0 days 00:07:43.780726                          
[2018-04-21 10:09:00.876752] INFO: exchange_blotter: poloniex order 444153990559 for eth_btc still open after 0 days 00:08:47.876616                           
[2018-04-21 10:09:01.325911] INFO: exchange_blotter: poloniex order 171699774378 for dash_btc still open after 0 days 00:08:43.325783                          
[2018-04-21 10:10:02.481622] INFO: exchange_blotter: poloniex order 444153990559 for eth_btc still open after 0 days 00:09:49.481463                           
[2018-04-21 10:10:02.679058] INFO: exchange_blotter: poloniex order 171699774378 for dash_btc still open after 0 days 00:09:44.678908
```
and so on. Catalyst never detects that these orders have been filled.

**Comment by AvishaiW** [comment]
Hi, can you please elaborate more on this issue? it seems from the logs that the filled orders are on the two pairs you brought up (`bch_btc` & `ltc_btc`) but the open orders are on different pairs (`etc_btc` & `zec_btc`)

**Comment by sam31415** [comment]
Hi, sorry for the confusion. I edited the logs to remove the lines unrelated to the problem. The problem in the first post is for the pairs etc_btc and zec_btc. I know the orders were filled by looking at the exchange. The algorithm detected that the balance for these pairs went down to zero (the two warning), but it didn't consider the orders as filled. As a result, it failed to add the proceeds of the sales into the cash account, creating an apparent loss. 

As far as the exchange is concerned, everything went smoothly. I see the orders in the history:
```
  | ZEC/BTC | Sell | Exchange | 0.03113212 | 0.02512898 | 0.00000196 BTC (0.25%) | 0.00078035 BTC | 2018-04-20 06:00:18
  | ZEC/BTC | Sell | Exchange | 0.03114000 | 0.01201778 | 0.00000094 BTC (0.25%) | 0.00037329 BTC | 2018-04-20 06:00:18
  | ZEC/BTC | Sell | Exchange | 0.03115000 | 0.01206398 | 0.00000094 BTC (0.25%) | 0.00037485 BTC | 2018-04-20 06:00:18
  | ZEC/BTC | Sell | Exchange | 0.03116000 | 0.01205813 | 0.00000094 BTC (0.25%) | 0.00037479 BTC | 2018-04-20 06:00:18
  | ZEC/BTC | Sell | Exchange | 0.03117000 | 0.01204436 | 0.00000094 BTC (0.25%) | 0.00037448 BTC | 2018-04-20 06:00:18
  | ZEC/BTC | Sell | Exchange | 0.03118000 | 0.01184430 | 0.00000092 BTC (0.25%) | 0.00036838 BTC | 2018-04-20 06:00:18
  | ETC/BTC | Sell | Exchange | 0.00219839 | 1.21984670 | 0.00000670 BTC (0.25%) | 0.00267499 BTC | 2018-04-20 06:00:16
```
The same is true for the issue reported in the second post. On the exchange everything looks normal:
```
  | DASH/BTC | Sell | Exchange | 0.05006652 | 0.03193609 | 0.00000400 BTC (0.25%) | 0.00159492 BTC | 2018-04-21 10:00:18
  | DASH/BTC | Sell | Exchange | 0.05010720 | 0.01816792 | 0.00000228 BTC (0.25%) | 0.00090806 BTC | 2018-04-21 10:00:18
  | ETH/BTC | Sell | Exchange | 0.06740368 | 0.03785017 | 0.00000638 BTC (0.25%) | 0.00254486 BTC | 2018-04-21 10:00:13
```

**Comment by manelrovira** [comment]
I have the same problem. I think it's because when retrieving the status of an order from the exhange the value of the filled field does not treat the decimals well. As an example these two orders were not closed properly:

`Order({'status': 0, 'broker_order_id': '1991960, 1991961', 'created': Timestamp('2018-04-18 02:17:15.307316+0000', tz='UTC'), 'limit_reached': True, 'stop': None, 'amount': 8.11, 'stop_reached': False, 'commission': 0.0010118
4, 'reason': None, 'limit': 0.00037294439999999997, 'sid': Trading Pair aion_btc(251436) Exchange: binance, Introduced On: 2009-01-01 00:00:00+00:00, Base Currency: aion, Quote Currency: btc, Exchange Leverage: 1.0, Minimum T
rade Size: 0.00999999977648 Last daily ingestion: None Last minutely ingestion: None, 'dt': Timestamp('2018-04-18 02:17:15.178000+0000', tz='UTC'), 'id': '10143107', 'filled': 8.110000000000001})
Order({'status': 0, 'broker_order_id': '5246279, 5246280, 5246281', 'created': Timestamp('2018-04-18 01:10:11.085832+0000', tz='UTC'), 'limit_reached': True, 'stop': None, 'amount': -1.44, 'stop_reached': False, 'commission':
 0.00098777, 'reason': None, 'limit': 0.002062866, 'sid': Trading Pair qtum_btc(861533) Exchange: binance, Introduced On: 2009-01-01 00:00:00+00:00, Base Currency: qtum, Quote Currency: btc, Exchange Leverage: 1.0, Minimum Tr
ade Size: 0.00999999977648 Last daily ingestion: None Last minutely ingestion: None, 'dt': Timestamp('2018-04-18 01:10:10.957000+0000', tz='UTC'), 'id': '21385899', 'filled': -1.4400000000000002})`

**Comment by sam31415** [comment]
@manelrovira That's interesting. It could also explain why sometimes a tiny balance remains after selling a position (of the order 1e-17 or so), as mentioned [here](https://catalyst.enigma.co/t/glitches-when-live-trading-on-poloniex/152).

**Comment by AvishaiW** [comment]
i'll try reproducing it, Thanks!

**Comment by AvishaiW** [comment]
Hi @trungtv, 
We are actively working on this issue for the past couple of days at high priority. It led us to a few issues that we are investigating. Hopefully we could introduce a complete solution shortly, will update when it's done. 

Of course, we want to find a proper solution which will solve this issue and won't harm any use cases.

**Comment by AvishaiW** [comment]
@sam31415 just to clarify, on which type of order does this issue occur? `order_target()`/`order()`/ all orders?

**Comment by sam31415** [comment]
I modified my algorithm in the mean time so I'm not sure whether it was `order_target()` or `order()`. Anyway all these functions are wrappers for a single ordering function, and I don't think the problem lies in the wrappers.

Actually my original post mentions it was with `order_target()`. But again, my guess is that the problem will occur no matter how you do the ordering.

**Comment by manelrovira** [comment]
This workaround have worked for me:  i've modified the definition of the property open_amount in catalyst/finance/order.py and all the orders have been filled since the modification. I know it's an ugly arrangement, but as long as the problem is not definitively fixed, it'll do.

`
    @property
    def open_amount(self):
        if abs(self.filled) > abs(self.amount):
            return 0.0
        if abs(self.filled - self.amount) < 0.000000001:
            return 0.0
        return self.amount - self.filled
`

**Comment by sam31415** [comment]
@manelrovira Thanks I'll try that.

@AvishaiW I should also mention that I did not encounter this problem on Binance. I had a few orders filled against 2 or 3 limit orders and catalyst handled them correctly (using the `order` function in this case).

**Comment by lenak25** [comment]
@manelrovira and @trungtv , did you also expedience this problem on Poloniex only (like @sam31415 indicated)?

**Comment by catalyst-support** [comment]
➤ Lena Kleyner commented:

testing

**Comment by manelrovira** [comment]
I'm having the problem in binance.

**Comment by lenak25** [comment]
Thanks for the update @manelrovira

**Comment by lenak25** [comment]
Fixed at release 0.5.10.
